File: language-support/js/daml-ledger/src/mapping/get_package_response.ts
// Copyright (c) 2019 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
// SPDX-License-Identifier: Apache-2.0

import * as mapping from '.';
import * as ledger from '..';
import * as grpc from 'daml-grpc';

export const GetPackageResponse: mapping.Mapping<grpc.GetPackageResponse, ledger.GetPackageResponse> = {
    toObject(response: grpc.GetPackageResponse): ledger.GetPackageResponse {
        return {
            hash: response.getHash(),
            hashFunction: ledger.HashFunction.SHA256,
            archivePayload: response.getArchivePayload_asB64()
        }
    },
    toMessage(response: ledger.GetPackageResponse): grpc.GetPackageResponse {
        const result = new grpc.GetPackageResponse();
        result.setHash(response.hash);
        result.setHashFunction(response.hashFunction);
        result.setArchivePayload(response.archivePayload);
        return result;
    }
}

File: sdk/canton/community/bindings-java/src/main/java/com/daml/ledger/javaapi/data/GetPackageResponse.java
// Copyright (c) 2024 Digital Asset (Switzerland) GmbH and/or its affiliates.
// Proprietary code. All rights reserved.

package com.daml.ledger.javaapi.data;

import com.google.protobuf.ByteString;
import java.util.EnumSet;
import java.util.Map;
import java.util.function.Function;
import java.util.stream.Collectors;
import org.checkerframework.checker.nullness.qual.NonNull;

public final class GetPackageResponse {

  // Clone of the PackageServiceOuterClass.HashFunction enumeration
  public enum HashFunction {
    SHA256(0),
    UNRECOGNIZED(-1),
    ;

    private final int value;

    private static Map<Integer, GetPackageResponse.HashFunction> valueToEnumMap =
        EnumSet.allOf(GetPackageResponse.HashFunction.class).stream()
            .collect(Collectors.toMap(e -> e.value, Function.identity()));

    private HashFunction(int value) {
      this.value = value;
    }

    public static GetPackageResponse.HashFunction valueOf(int value) {
      return valueToEnumMap.getOrDefault(value, UNRECOGNIZED);
    }
  }

  private final HashFunction hashFunction;
  private final String hash;
  private final ByteString archivePayload;

  public GetPackageResponse(
      HashFunction hashFunction, @NonNull String hash, @NonNull ByteString archivePayload) {
    this.hashFunction = hashFunction;
    this.hash = hash;
    this.archivePayload = archivePayload;
  }

  public HashFunction getHashFunction() {
    return hashFunction;
  }

  public String getHash() {
    return hash;
  }

  public byte[] getArchivePayload() {
    return archivePayload.toByteArray();
  }

  public static GetPackageResponse fromProto(
      com.daml.ledger.api.v1.PackageServiceOuterClass.GetPackageResponse p) {
    return new GetPackageResponse(
        HashFunction.valueOf(p.getHashFunctionValue()), p.getHash(), p.getArchivePayload());
  }
}
